File: py_solver/solver/distribution.py
"""
Implements the game's Distribution System (see docs/game-logic.md):
FairShare followed by Augmenting Repair, run identically for the heat
instance (Reactor -> Generator) and the cooling instance
(Cooler -> Generator/DirectProducer).

The per-supplier / per-consumer *split* is ported from Unity
`FlowNetwork.cs` (FairShare + Edmonds-Karp repair). Total max-flow is
unchanged; the split is what decides which generators clear cooling.

FairShare is simulated literally, in the documented spatial order
(ascending X, then descending Y), including its bounded round loop and the
even re-split of each supplier's leftover. Rounds are budgeted per connected
component of the supplier<->consumer graph -- Unity already solves each
FlowNetwork island separately, so `maxRounds = s` on an island is the same
budget.

Augmenting Repair is Unity's BFS: source enqueues every supplier with
remaining supply; suppliers enqueue every unvisited adjacent consumer
(forward residual unbounded, no jump-to-sink); a consumer with leftover
demand connects to the sink and ends that search, otherwise it walks
back-edges. Node space matches FlowNetwork: suppliers, then consumers,
then source, then sink.
"""

from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple

from solver.constants import EPS, spatial_key
from solver.rules import TileId, is_adjacent


@dataclass
class Node:
    """One endpoint of a distribution graph: a tile with a capacity to give or absorb."""

    id: TileId
    capacity: float


def run_distribution(
    suppliers: List[Node],
    consumers: List[Node],
    neighbor_map: Optional[Dict[TileId, List[TileId]]] = None,
) -> Tuple[Dict[TileId, float], Dict[TileId, float]]:
    """Returns (supplier_id -> total sent, consumer_id -> total received)."""
    if not suppliers or not consumers:
        return {s.id: 0.0 for s in suppliers}, {c.id: 0.0 for c in consumers}

    sorted_suppliers = sorted(suppliers, key=lambda s: spatial_key(*s.id))
    sorted_consumers = sorted(consumers, key=lambda c: spatial_key(*c.id))

    num_suppliers = len(sorted_suppliers)
    num_consumers = len(sorted_consumers)

    consumer_id_to_idx = {c.id: j for j, c in enumerate(sorted_consumers)}

    # Flattened bipartite edges, matching FlowNetwork.BuildIsland:
    # suppliers in spatial order, each supplier's consumer neighbours in
    # ascending local consumer index.
    edge_supplier: List[int] = []
    edge_consumer: List[int] = []
    supplier_edges: List[List[int]] = [[] for _ in range(num_suppliers)]
    consumer_edges: List[List[int]] = [[] for _ in range(num_consumers)]

    for si, s in enumerate(sorted_suppliers):
        if neighbor_map is not None:
            locals_ci = [
                consumer_id_to_idx[n]
                for n in neighbor_map.get(s.id, ())
                if n in consumer_id_to_idx
            ]
            locals_ci.sort()
        else:
            locals_ci = [
                ci
                for ci, c in enumerate(sorted_consumers)
                if is_adjacent(s.id, c.id)
            ]
        for ci in locals_ci:
            edge = len(edge_supplier)
            edge_supplier.append(si)
            edge_consumer.append(ci)
            supplier_edges[si].append(edge)
            consumer_edges[ci].append(edge)

    cap = [s.capacity for s in sorted_suppliers]
    demand = [c.capacity for c in sorted_consumers]
    sup_remaining = list(cap)
    dem_remaining = list(demand)
    edge_flow = [0.0] * len(edge_supplier)

    # --- Phase 1: FairShare (FlowNetwork.FairShare inner loop) ---
    # Each round walks every supplier in spatial order and splits whatever that
    # supplier still holds EVENLY across its adjacent consumers that still have
    # demand. A leftover (created when an even share overshoots a consumer's
    # remaining capacity) is re-split evenly on the NEXT round -- never handed
    # whole to one neighbour. The round loop sits OUTSIDE the supplier loop, so
    # a supplier's round-N split sees what the others did in round N-1.
    #
    # A supplier gets as many rounds as there are suppliers in **its own
    # connected component** of the supplier<->consumer graph -- not on the
    # island, and not on the board. Buildings that cannot exchange anything do
    # not lengthen each other's redistribution. Measured in-game: an isolated
    # reactor keeps its single round even with other reactors elsewhere on the
    # same island, and the same layout on an island of its own behaves
    # identically. See docs/game-logic.md.
    components: Optional[List[List[int]]] = None

    def get_components() -> List[List[int]]:
        nonlocal components
        if components is None:
            components = _supplier_components(
                num_suppliers, supplier_edges, edge_consumer, consumer_edges, edge_supplier
            )
        return components

    def fairshare_round(bound: Optional[List[int]], round_idx: int) -> float:
        moved = 0.0
        for si in range(num_suppliers):
            if bound is not None and round_idx >= bound[si]:
                continue
            if sup_remaining[si] <= EPS:
                continue

            edges = supplier_edges[si]
            needy = 0
            for edge in edges:
                if dem_remaining[edge_consumer[edge]] > EPS:
                    needy += 1
            if needy == 0:
                continue

            share = sup_remaining[si] / needy
            for edge in edges:
                if sup_remaining[si] <= EPS:
                    break
                ci = edge_consumer[edge]
                if dem_remaining[ci] <= EPS:
                    continue
                give = share
                if dem_remaining[ci] < give:
                    give = dem_remaining[ci]
                if sup_remaining[si] < give:
                    give = sup_remaining[si]
                if give <= EPS:
                    continue
                edge_flow[edge] += give
                sup_remaining[si] -= give
                dem_remaining[ci] -= give
                moved += give
        return moved

    # Round 1 needs no component map: every supplier's component contains at
    # least itself, so every supplier is entitled to one round.
    fairshare_round(None, 0)

    # Later rounds need the per-supplier bound, which costs a graph walk. Skip
    # it unless some supplier could actually still move something -- the common
    # case settles in one round and never pays for the decomposition.
    if any(
        sup_remaining[si] > EPS
        and any(dem_remaining[edge_consumer[e]] > EPS for e in supplier_edges[si])
        for si in range(num_suppliers)
    ):
        supplier_round_budget = [0] * num_suppliers
        for members in get_components():
            n = len(members)
            for si in members:
                supplier_round_budget[si] = n
        for round_idx in range(1, max(supplier_round_budget, default=1)):
            if fairshare_round(supplier_round_budget, round_idx) <= EPS:
                # A round that moves nothing leaves the state untouched, so no
                # later round can move anything either.
                break

    # --- Phase 2: Augmenting Repair (FlowNetwork.Augment) ---
    # Node space matches Unity:
    #   [0 .. S-1]       suppliers
    #   [S .. S+C-1]     consumers
    #   S+C              source
    #   S+C+1            sink
    # Augmenting Repair can only raise total delivery, so it needs BOTH a
    # supplier with something left to give and a consumer with room to take it.
    # Whenever either side is exhausted -- which is common, since layouts tend
    # to be tight on one of the two -- the whole phase is a guaranteed no-op and
    # its BFS machinery can be skipped entirely.
    if not any(r > EPS for r in sup_remaining) or not any(r > EPS for r in dem_remaining):
        return (
            {s.id: cap[i] - sup_remaining[i] for i, s in enumerate(sorted_suppliers)},
            {c.id: demand[j] - dem_remaining[j] for j, c in enumerate(sorted_consumers)},
        )

    source = num_suppliers + num_consumers
    sink = source + 1
    total_nodes = sink + 1

    visited_token = [0] * total_nodes
    parent_node = [0] * total_nodes
    via_edge = [0] * total_nodes
    queue = [0] * total_nodes
    token = 0

    # Suppliers grouped into connected components of the supplier<->consumer
    # graph. An augmenting path can never leave its component, and augmenting
    # inside one component cannot change any other's residual state, so each
    # can be repaired independently -- identical result, but every BFS scans
    # one component instead of the whole island. On a solved map-1 layout the
    # heat graph splits in two and the cooling graph into five.
    #
    # Each component's members MUST stay in ascending index order. The BFS
    # expands suppliers in that order, and when several augmenting paths are
    # equally short the first one reached wins -- so discovery order here would
    # silently pick a different (equally maximal, but differently split) flow,
    # and the split is what decides which buildings clear their cooling.
    for source_suppliers in get_components():
        while True:
            if not any(sup_remaining[si] > EPS for si in source_suppliers):
                break

            token += 1
            visited_token[source] = token
            parent_node[source] = -1
            queue[0] = source
            q_head = 0
            q_tail = 1
            reached_sink = False

            while q_head < q_tail and not reached_sink:
                node = queue[q_head]
                q_head += 1

                if node == source:
                    for si in source_suppliers:
                        if visited_token[si] == token:
                            continue
                        if sup_remaining[si] > EPS:
                            visited_token[si] = token
                            parent_node[si] = source
                            via_edge[si] = -1
                            queue[q_tail] = si
                            q_tail += 1

                elif node < num_suppliers:
                    # Supplier -> every unvisited adjacent consumer. Forward
                    # residual is unbounded. Do not jump to the sink here:
                    # FlowNetwork enqueues the consumer and lets the consumer
                    # node connect to the sink, so later same-layer suppliers
                    # still expand. Jumping from the supplier marks later
                    # consumers unvisited and can pick a different equal-length
                    # path.
                    for edge in supplier_edges[node]:
                        con_node = num_suppliers + edge_consumer[edge]
                        if visited_token[con_node] == token:
                            continue
                        visited_token[con_node] = token
                        parent_node[con_node] = node
                        via_edge[con_node] = edge
                        queue[q_tail] = con_node
                        q_tail += 1

                elif node < source:
                    ci = node - num_suppliers
                    if visited_token[sink] != token and dem_remaining[ci] > EPS:
                        visited_token[sink] = token
                        parent_node[sink] = node
                        via_edge[sink] = -1
                        reached_sink = True
                        break

                    for edge in consumer_edges[ci]:
                        if edge_flow[edge] <= EPS:
                            continue
                        sup_node = edge_supplier[edge]
                        if visited_token[sup_node] == token:
                            continue
                        visited_token[sup_node] = token
                        parent_node[sup_node] = node
                        via_edge[sup_node] = edge
                        queue[q_tail] = sup_node
                        q_tail += 1

            if not reached_sink:
                break

            bottleneck = float("inf")
            node = sink
            while node != source:
                prev = parent_node[node]
                if node == sink:
                    residual = dem_remaining[prev - num_suppliers]
                elif prev == source:
                    residual = sup_remaining[node]
                elif node < num_suppliers:
                    residual = edge_flow[via_edge[node]]
                else:
                    residual = float("inf")
                if residual < bottleneck:
                    bottleneck = residual
                node = prev

            if bottleneck == float("inf") or bottleneck <= EPS:
                break

            node = sink
            while node != source:
                prev = parent_node[node]
                if node == sink:
                    dem_remaining[prev - num_suppliers] -= bottleneck
                elif prev == source:
                    sup_remaining[node] -= bottleneck
                elif node < num_suppliers:
                    edge_flow[via_edge[node]] -= bottleneck
                else:
                    edge_flow[via_edge[node]] += bottleneck
                node = prev

    return (
        {s.id: cap[i] - sup_remaining[i] for i, s in enumerate(sorted_suppliers)},
        {c.id: demand[j] - dem_remaining[j] for j, c in enumerate(sorted_consumers)},
    )


def _supplier_components(
    num_suppliers: int,
    supplier_edges: List[List[int]],
    edge_consumer: List[int],
    consumer_edges: List[List[int]],
    edge_supplier: List[int],
) -> List[List[int]]:
    """
    Connected components of the supplier<->consumer graph, each as a list of
    supplier indices in ascending order.

    That count is the supplier's FairShare round budget: two clusters that
    share no consumer cannot affect each other, so neither may lengthen the
    other's redistribution. Repair walks the same components, and must keep
    this ascending-index order -- discovery order silently picks a different
    equally-short augmenting path.
    """
    component_of = [-1] * num_suppliers
    components: List[List[int]] = []
    for start in range(num_suppliers):
        if component_of[start] != -1:
            continue
        members = [start]
        component_of[start] = len(components)
        seen_consumers = set()
        stack = [start]
        while stack:
            si = stack.pop()
            for edge in supplier_edges[si]:
                ci = edge_consumer[edge]
                if ci in seen_consumers:
                    continue
                seen_consumers.add(ci)
                for other_edge in consumer_edges[ci]:
                    other = edge_supplier[other_edge]
                    if component_of[other] == -1:
                        component_of[other] = len(components)
                        members.append(other)
                        stack.append(other)
        members.sort()
        components.append(members)
    return components

File: packages/solver/src/solver/distribution.ts
import { EPS } from "./constants";
import type { IslandContext } from "./context";

/**
 * Implements the game's Distribution System: FairShare followed by Augmenting
 * Repair, run identically for the heat instance (Reactor -> Generator) and the
 * cooling instance (Cooler -> Generator/DirectProducer). Port of
 * Unity's `FlowNetwork.cs`.
 *
 * FairShare is simulated literally, in the documented spatial order (ascending
 * X, then descending Y), including its bounded round loop and the even
 * re-split of each supplier's leftover.
 *
 * Augmenting Repair is Unity's BFS: the source enqueues every supplier with
 * remaining supply; a supplier enqueues every unvisited adjacent consumer
 * (forward residual unbounded); a consumer with leftover demand connects to the
 * sink and ends that search, otherwise it walks back-edges.
 *
 * Two shapes here are load-bearing and neither is arbitrary:
 *
 * - A supplier does **not** jump straight to the sink on finding a consumer
 *   with room. It enqueues the consumer and lets the consumer node connect,
 *   because the shortcut marks later consumers visited and can select a
 *   different equal-length augmenting path — same maximum flow, different
 *   split, and the split is what decides which buildings clear their cooling.
 * - Supply and demand are counted DOWN. `cap - sent` accumulated separately is
 *   not bit-identical to a decremented remainder, and the parity fixtures are
 *   asserted exactly.
 *
 * The node numbering matches `FlowNetwork`: suppliers, then consumers, then
 * source, then sink.
 */

/**
 * Distributes `ctx.dist.supplierCap` over `ctx.dist.consumerCap`, writing the
 * outcome to `ctx.dist.supplierSent` / `ctx.dist.consumerReceived` (aligned
 * with `supplierTiles` / `consumerTiles`).
 *
 * Both tile lists MUST be in ascending tile-index order, which for an
 * `IslandContext` is exactly the game's spatial processing order — so unlike
 * ascending tile index, this never has to sort.
 */
export function runDistribution(
  supplierTiles: Int32Array,
  numSuppliers: number,
  consumerTiles: Int32Array,
  numConsumers: number,
  ctx: IslandContext,
): void {
  const d = ctx.dist;
  const supplierCap = d.supplierCap;
  const consumerCap = d.consumerCap;
  const supplierSent = d.supplierSent;
  const consumerReceived = d.consumerReceived;

  supplierSent.fill(0, 0, numSuppliers);
  consumerReceived.fill(0, 0, numConsumers);

  if (numSuppliers === 0 || numConsumers === 0) return;

  const supRem = d.remSupplierCap;
  const demRem = d.remConsumerCap;
  for (let i = 0; i < numSuppliers; i++) supRem[i] = supplierCap[i];
  for (let j = 0; j < numConsumers; j++) demRem[j] = consumerCap[j];

  const finish = (): void => {
    for (let i = 0; i < numSuppliers; i++)
      supplierSent[i] = supplierCap[i] - supRem[i];
    for (let j = 0; j < numConsumers; j++) {
      consumerReceived[j] = consumerCap[j] - demRem[j];
    }
  };

  const stamp = ++d.stamp;

  // Map consumer tiles to their consumer index so adjacency can be filtered
  // with an O(1) lookup instead of a pairwise `isAdjacent` scan.
  for (let j = 0; j < numConsumers; j++) {
    const tile = consumerTiles[j];
    d.conIdxOf[tile] = j;
    d.conIdxStamp[tile] = stamp;
  }

  // Supplier -> consumer adjacency, CSR-packed. Neighbor lists are already in
  // ascending index order, so the consumer indices come out spatially sorted —
  // the same order `FlowNetwork.BuildIsland` gives each supplier's edge list.
  // The reverse direction is only walked by Augmenting Repair, which often
  // turns out to have nothing to do, so it is built on demand.
  const supAdjStart = d.supAdjStart;
  const supAdjItems = d.supAdjItems;
  let write = 0;
  for (let i = 0; i < numSuppliers; i++) {
    supAdjStart[i] = write;
    const nb = ctx.neighbors[supplierTiles[i]];
    for (let k = 0; k < nb.length; k++) {
      const tile = nb[k];
      if (d.conIdxStamp[tile] === stamp)
        supAdjItems[write++] = d.conIdxOf[tile];
    }
  }
  supAdjStart[numSuppliers] = write;

  const flow = d.flow;
  flow.fill(0, 0, numSuppliers * numConsumers);

  // Consumer->supplier adjacency and the component decomposition are needed by
  // FairShare only when it runs past its first round, and by Augmenting Repair
  // only when it has something to do. Both are common enough to skip and
  // expensive enough to matter, so they are built at most once, on demand.
  let conAdjBuilt = false;
  let componentsBuilt = false;
  let numComponents = 0;

  // --- Phase 1: FairShare (FlowNetwork.FairShare inner loop) ---
  // Each round walks every supplier in spatial order and splits whatever that
  // supplier still holds EVENLY across its adjacent consumers that still have
  // demand. A leftover (created when an even share overshoots a consumer's
  // remaining capacity) is re-split evenly on the NEXT round -- never handed
  // whole to one neighbour. The round loop sits OUTSIDE the supplier loop, so a
  // supplier's round-N split sees what the others did in round N-1.
  //
  // A supplier gets as many rounds as there are suppliers in **its own
  // connected component** of the supplier<->consumer graph -- not on the
  // island, and not on the board. Buildings that cannot exchange anything must
  // not lengthen each other's redistribution. All three properties are measured
  // in-game; see `distribution.test.ts`.
  const supRoundBudget = d.supRoundBudget;

  /** Returns the total mass moved, which is how the round loop knows to stop. */
  const runRound = (budgeted: boolean, roundIdx: number): number => {
    let moved = 0.0;

    for (let sIdx = 0; sIdx < numSuppliers; sIdx++) {
      if (budgeted && roundIdx >= supRoundBudget[sIdx]) continue;
      if (supRem[sIdx] <= EPS) continue;

      const from = supAdjStart[sIdx];
      const to = supAdjStart[sIdx + 1];

      let needy = 0;
      for (let k = from; k < to; k++) {
        if (demRem[supAdjItems[k]] > EPS) needy++;
      }
      if (needy === 0) continue;

      const share = supRem[sIdx] / needy;
      const rowBase = sIdx * numConsumers;

      for (let k = from; k < to; k++) {
        if (supRem[sIdx] <= EPS) break;
        const cIdx = supAdjItems[k];
        if (demRem[cIdx] <= EPS) continue;
        let give = share;
        if (demRem[cIdx] < give) give = demRem[cIdx];
        if (supRem[sIdx] < give) give = supRem[sIdx];
        if (give <= EPS) continue;
        flow[rowBase + cIdx] += give;
        supRem[sIdx] -= give;
        demRem[cIdx] -= give;
        moved += give;
      }
    }

    return moved;
  };

  // Round 1 needs no component map: every supplier's component contains at
  // least itself, so every supplier is entitled to one round.
  runRound(false, 0);

  // Every later round needs the per-supplier budget, and that costs a graph
  // walk. Most layouts settle in one round, so only pay for it when some
  // supplier could still move something.
  let needsMoreRounds = false;
  for (let i = 0; i < numSuppliers && !needsMoreRounds; i++) {
    if (supRem[i] <= EPS) continue;
    for (let k = supAdjStart[i]; k < supAdjStart[i + 1]; k++) {
      if (demRem[supAdjItems[k]] > EPS) {
        needsMoreRounds = true;
        break;
      }
    }
  }

  if (needsMoreRounds) {
    buildConsumerAdjacency(
      supplierTiles,
      numSuppliers,
      consumerTiles,
      numConsumers,
      ctx,
      stamp,
    );
    conAdjBuilt = true;
    numComponents = buildComponents(numSuppliers, numConsumers, d);
    componentsBuilt = true;

    let maxRounds = 1;
    for (let i = 0; i < numSuppliers; i++) {
      if (supRoundBudget[i] > maxRounds) maxRounds = supRoundBudget[i];
    }
    for (let roundIdx = 1; roundIdx < maxRounds; roundIdx++) {
      // A round that moves nothing leaves the state untouched, so no later
      // round can move anything either.
      if (runRound(true, roundIdx) <= EPS) break;
    }
  }

  // --- Phase 2: Augmenting Repair (FlowNetwork.Augment) ---
  // Augmenting Repair can only raise total delivery, so it needs BOTH a
  // supplier with something left to give and a consumer with room to take it.
  // Whenever either side is exhausted -- which is common, since layouts tend to
  // be tight on one of the two -- the whole phase is a guaranteed no-op and its
  // BFS machinery can be skipped entirely.
  let anySupplierLeft = false;
  for (let i = 0; i < numSuppliers; i++) {
    if (supRem[i] > EPS) {
      anySupplierLeft = true;
      break;
    }
  }
  if (!anySupplierLeft) {
    finish();
    return;
  }

  let anyConsumerRoom = false;
  for (let j = 0; j < numConsumers; j++) {
    if (demRem[j] > EPS) {
      anyConsumerRoom = true;
      break;
    }
  }
  if (!anyConsumerRoom) {
    finish();
    return;
  }

  // Consumer -> supplier adjacency. FairShare already builds this whenever it
  // needed a second round, so re-do it only if that did not happen.
  if (!conAdjBuilt) {
    buildConsumerAdjacency(
      supplierTiles,
      numSuppliers,
      consumerTiles,
      numConsumers,
      ctx,
      stamp,
    );
  }
  const conAdjStart = d.conAdjStart;
  const conAdjItems = d.conAdjItems;

  const source = numSuppliers + numConsumers;
  const sink = source + 1;
  const visitedToken = d.visitedToken;
  const parentNode = d.parentNode;
  const queue = d.queue;
  visitedToken.fill(0, 0, sink + 1);
  let token = 0;

  // Suppliers grouped into connected components of the supplier<->consumer
  // graph. An augmenting path can never leave its component, and augmenting
  // inside one cannot change any other's residual state, so each is repaired
  // independently -- identical result, but every BFS scans one component
  // instead of the whole island. On a solved map-1 layout the heat graph splits
  // in two and the cooling graph into five.
  //
  // FairShare already needs this decomposition whenever it runs a second round
  // (a supplier's round budget is its component's supplier count), so reuse it
  // when it is already there.
  const componentMembers = d.componentMembers;
  const componentStart = d.componentStart;
  if (!componentsBuilt) {
    numComponents = buildComponents(numSuppliers, numConsumers, d);
  }

  for (let comp = 0; comp < numComponents; comp++) {
    const memberFrom = componentStart[comp];
    const memberTo = componentStart[comp + 1];

    for (;;) {
      let anyLeft = false;
      for (let m = memberFrom; m < memberTo; m++) {
        if (supRem[componentMembers[m]] > EPS) {
          anyLeft = true;
          break;
        }
      }
      if (!anyLeft) break;

      token += 1;
      visitedToken[source] = token;
      queue[0] = source;
      let qHead = 0;
      let qTail = 1;
      let reachedSink = false;

      while (qHead < qTail && !reachedSink) {
        const node = queue[qHead++];

        if (node === source) {
          for (let m = memberFrom; m < memberTo; m++) {
            const sIdx = componentMembers[m];
            if (visitedToken[sIdx] === token) continue;
            if (supRem[sIdx] > EPS) {
              visitedToken[sIdx] = token;
              parentNode[sIdx] = source;
              queue[qTail++] = sIdx;
            }
          }
        } else if (node < numSuppliers) {
          // Supplier -> every unvisited adjacent consumer. Forward residual is
          // unbounded, and the sink is deliberately NOT reached from here.
          for (let k = supAdjStart[node]; k < supAdjStart[node + 1]; k++) {
            const conNode = numSuppliers + supAdjItems[k];
            if (visitedToken[conNode] === token) continue;
            visitedToken[conNode] = token;
            parentNode[conNode] = node;
            queue[qTail++] = conNode;
          }
        } else if (node < source) {
          const cIdx = node - numSuppliers;
          if (visitedToken[sink] !== token && demRem[cIdx] > EPS) {
            visitedToken[sink] = token;
            parentNode[sink] = node;
            reachedSink = true;
            break;
          }

          for (let m = conAdjStart[cIdx]; m < conAdjStart[cIdx + 1]; m++) {
            const sIdx = conAdjItems[m];
            if (flow[sIdx * numConsumers + cIdx] <= EPS) continue;
            if (visitedToken[sIdx] === token) continue;
            visitedToken[sIdx] = token;
            parentNode[sIdx] = node;
            queue[qTail++] = sIdx;
          }
        }
      }

      if (!reachedSink) break;

      // Walk back from SINK to SOURCE to find the path bottleneck. A forward
      // edge (consumer reached from a supplier) has unbounded residual; only
      // the source edge, the sink edge and back-edges constrain the path.
      let bottleneck = Infinity;
      for (let node = sink; node !== source;) {
        const prev = parentNode[node];
        let residual: number;
        if (node === sink) {
          residual = demRem[prev - numSuppliers];
        } else if (prev === source) {
          residual = supRem[node];
        } else if (node < numSuppliers) {
          residual = flow[node * numConsumers + (prev - numSuppliers)];
        } else {
          residual = Infinity;
        }
        if (residual < bottleneck) bottleneck = residual;
        node = prev;
      }

      if (bottleneck === Infinity || bottleneck <= EPS) break;

      for (let node = sink; node !== source;) {
        const prev = parentNode[node];
        if (node === sink) {
          demRem[prev - numSuppliers] -= bottleneck;
        } else if (prev === source) {
          supRem[node] -= bottleneck;
        } else if (node < numSuppliers) {
          flow[node * numConsumers + (prev - numSuppliers)] -= bottleneck;
        } else {
          flow[prev * numConsumers + (node - numSuppliers)] += bottleneck;
        }
        node = prev;
      }
    }
  }

  finish();
}

/**
 * Fills `d.conAdjStart` / `d.conAdjItems` with each consumer's adjacent
 * suppliers, in ascending supplier index (i.e. spatial) order.
 */
function buildConsumerAdjacency(
  supplierTiles: Int32Array,
  numSuppliers: number,
  consumerTiles: Int32Array,
  numConsumers: number,
  ctx: IslandContext,
  stamp: number,
): void {
  const d = ctx.dist;
  for (let i = 0; i < numSuppliers; i++) {
    const tile = supplierTiles[i];
    d.supIdxOf[tile] = i;
    d.supIdxStamp[tile] = stamp;
  }
  const conAdjStart = d.conAdjStart;
  const conAdjItems = d.conAdjItems;
  let write = 0;
  for (let j = 0; j < numConsumers; j++) {
    conAdjStart[j] = write;
    const nb = ctx.neighbors[consumerTiles[j]];
    for (let k = 0; k < nb.length; k++) {
      const tile = nb[k];
      if (d.supIdxStamp[tile] === stamp)
        conAdjItems[write++] = d.supIdxOf[tile];
    }
  }
  conAdjStart[numConsumers] = write;
}

/**
 * Groups suppliers into connected components of the supplier<->consumer graph,
 * filling `d.componentOf` / `componentMembers` / `componentStart`, and writes
 * each supplier's component size into `d.supRoundBudget` — which is that
 * supplier's FairShare round budget. Returns the component count.
 *
 * Requires `buildConsumerAdjacency` to have run. Each component's members MUST
 * stay in ascending index order: the repair BFS expands suppliers in that
 * order, and when several augmenting paths are equally short the first one
 * reached wins, so discovery order would silently pick a different (equally
 * maximal, but differently split) flow — and the split decides which buildings
 * clear their cooling threshold.
 */
function buildComponents(
  numSuppliers: number,
  numConsumers: number,
  d: IslandContext["dist"],
): number {
  const componentOf = d.componentOf;
  const componentMembers = d.componentMembers;
  const componentStart = d.componentStart;
  const supRoundBudget = d.supRoundBudget;
  const conSeen = d.conSeen;
  const stack = d.stack;
  const supAdjStart = d.supAdjStart;
  const supAdjItems = d.supAdjItems;
  const conAdjStart = d.conAdjStart;
  const conAdjItems = d.conAdjItems;

  componentOf.fill(-1, 0, numSuppliers);
  conSeen.fill(0, 0, numConsumers);

  let numComponents = 0;
  let memberWrite = 0;
  for (let start = 0; start < numSuppliers; start++) {
    if (componentOf[start] !== -1) continue;

    const componentBegin = memberWrite;
    componentStart[numComponents] = componentBegin;
    componentOf[start] = numComponents;
    componentMembers[memberWrite++] = start;

    let stackTop = 0;
    stack[stackTop++] = start;
    while (stackTop > 0) {
      const sIdx = stack[--stackTop];
      for (let k = supAdjStart[sIdx]; k < supAdjStart[sIdx + 1]; k++) {
        const cIdx = supAdjItems[k];
        if (conSeen[cIdx] === 1) continue;
        conSeen[cIdx] = 1;
        for (let m = conAdjStart[cIdx]; m < conAdjStart[cIdx + 1]; m++) {
          const other = conAdjItems[m];
          if (componentOf[other] === -1) {
            componentOf[other] = numComponents;
            componentMembers[memberWrite++] = other;
            stack[stackTop++] = other;
          }
        }
      }
    }

    componentMembers.subarray(componentBegin, memberWrite).sort();
    const size = memberWrite - componentBegin;
    for (let m = componentBegin; m < memberWrite; m++) {
      supRoundBudget[componentMembers[m]] = size;
    }
    numComponents++;
  }
  componentStart[numComponents] = memberWrite;
  return numComponents;
}
